src/SFML/Graphics/Shader.cpp

```diff
@@ -156,11 +210,50 @@ Shader::CurrentTextureType Shader::CurrentTexture;
 
 
 ////////////////////////////////////////////////////////////
+struct Shader::UniformBinder : private sf::NonCopyable
+{
+    ////////////////////////////////////////////////////////////
+    /// \brief Constructor: set up state before uniform is set
+    ///
+    ////////////////////////////////////////////////////////////
+    UniformBinder(Shader* shader, const std::string& name) :
+    program(0),
+    location(-1)
+    {
+        if (shader->m_shaderProgram)
```

Comment:
> The constructor and destructor in GlResource are there for a reason: to make sure that the shared context is available before anybody calls ensureContext. We already had 2 cases of bugs which involved ensureContext being called before any GlResource was constructed

But this can't be the case here, since we're already working in a GL resource (```sf::Shader```), which has done the necessary job in its own constructor.

> The constructor and destructor don't even do that much. If the shared context already exists, they simply lock and unlock a mutex.

Still better to not call them. And who knows how the ```GlResource``` class will evolve in the future, especially after all the work you're doing about context stuff.